pkg/webhook/certs.go

```diff
@@ -0,0 +1,369 @@
+	// explicitly rotate on the first round so that the certificate
+	// can be bootstrapped, otherwise manager exits before a cert can be written
+	crLog.Info("starting cert rotator controller")
+	defer crLog.Info("stopping cert rotator controller")
+	if restart, err := cr.refreshCertIfNeeded(); err != nil {
+		crLog.Error(err, "could not refresh cert on startup")
+		return errors.Wrap(err, "could not refresh cert on startup")
+	} else if restart {
+		crLog.Info("certs refreshed, restarting server")
+		return nil
+	}
+	ticker := time.NewTicker(12 * time.Hour)
+	done := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				if restart, err := cr.refreshCertIfNeeded(); err != nil {
+					crLog.Error(err, "error rotating certs")
+				} else if restart {
+					crLog.Info("certs refreshed, restarting server")
+					close(done)
+					return
+				}
+			case <-stop:
+				close(done)
+				return
+			}
+		}
+	}()
+
+	<-done
+	ticker.Stop()
+	return nil
+}
+
+// refreshCertIfNeeded returns whether the cert was refreshed and any errors
+func (cr *certRotator) refreshCertIfNeeded() (bool, error) {
+	secret := &corev1.Secret{}
+	if err := cr.client.Get(context.Background(), secretKey, secret); err != nil {
+		return false, errors.Wrap(err, "acquiring secret to update certificates")
+	}
+	if secret.Data == nil || !validCACert(secret.Data[caCertName], secret.Data[caKeyName]) {
+		crLog.Info("refreshing CA and server certs")
+		return cr.refreshCerts(true, secret)
+	}
+	if !validServerCert(secret.Data[caCertName], secret.Data[certName], secret.Data[keyName]) {
+		crLog.Info("refreshing server certs")
+		return cr.refreshCerts(false, secret)
+	}
+	crLog.Info("no cert refresh needed")
+	return false, nil
+}
+
+func (cr *certRotator) refreshCerts(refreshCA bool, secret *corev1.Secret) (bool, error) {
+	var caArtifacts *KeyPairArtifacts
+	if refreshCA {
+		var err error
+		caArtifacts, err = createCACert()
+		if err != nil {
+			return false, err
+		}
+	} else {
+		var err error
+		caArtifacts, err = buildArtifactsFromSecret(secret)
+		if err != nil {
+			return false, err
+		}
+	}
+	cert, key, err := createCertPEM(caArtifacts)
+	if err != nil {
+		return false, err
+	}
+	// execute writeWebhookConfig first because update is triggered when the
+	// secret is invalid. Therefore writing the secret last means that this
+	// process will be re-triggered if any update fails.
+	if err := cr.writeWebhookConfig(caArtifacts.CertPEM); err != nil {
+		return false, err
+	}
+	if err := cr.writeSecret(cert, key, caArtifacts, secret); err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+func (cr *certRotator) writeWebhookConfig(certPem []byte) error {
+	vwh := &unstructured.Unstructured{}
+	vwh.SetGroupVersionKind(schema.GroupVersionKind{Group: "admissionregistration.k8s.io", Version: "v1beta1", Kind: "ValidatingWebhookConfiguration"})
+	vwhKey := types.NamespacedName{Name: "gatekeeper-validating-webhook-configuration"}
```

Comment:
What is the reason for changing the name of this webhookconfig to `gatekeeper-validating-webhook-configuration`

Reply:
Per our conversation this is for compatibility with `controller-gen`